sim/dex.ts

```diff
@@ -506,28 +505,21 @@ export class ModdedDex {
 			this.includeFormats();
 		}
 		for (const dataType of DATA_TYPES.concat('Aliases')) {
-			const BattleData = this.loadDataFile(basePath, dataType);
-			if (BattleData !== dataCache[dataType]) dataCache[dataType] = Object.assign(BattleData, dataCache[dataType]);
+			dataCache[dataType] = this.loadDataFile(basePath, dataType);
 			if (dataType === 'Rulesets' && !parentDex) {
 				for (const format of this.formats.all()) {
-					BattleData[format.id] = {...format, ruleTable: null};
+					dataCache.Rulesets[format.id] = {...format, ruleTable: null};
 				}
 			}
 		}
 		if (parentDex) {
 			for (const dataType of DATA_TYPES) {
 				const parentTypedData: DexTable<any> = parentDex.data[dataType];
-				const childTypedData: DexTable<any> = dataCache[dataType] || (dataCache[dataType] = {});
-				// if child is empty and there's no Scripts.init, there's no need to copy, just re-use.
-				let childIsEmpty = true;
-				for (const k in childTypedData) { // eslint-disable-line @typescript-eslint/no-unused-vars
-					childIsEmpty = false;
-					break;
-				}
-				if (dataType !== 'Pokedex' && childIsEmpty && !Scripts.init) {
+				if (!dataCache[dataType] && !Scripts.init) {
```

Comment:
I think we still need the !== 'Pokedex' check, as counter-intuitive as that is.

When I add the deepFreeze line below,
```js
                // Execute initialization script.
                if (Scripts.init) Scripts.init.call(this);
                this.deepFreeze(dataCache);
```
npm test fails with
```
  2116 passing (10s)
  87 pending
  1 failing

  1) Burn
       should reduce atk to 50% of its original value in Stadium:
     TypeError: Cannot read properties of null (reading 'getStat')
      at Context.<anonymous> (test/sim/misc/statuses.js:41:38)
      at processImmediate (node:internal/timers:476:21)
```

When I add `Dex.includeMods()` and `Dex.includeModData()` to the very bottom of `./sim/dex.ts`, + logging, I get this
```
> mocha

loading data for fullpotential
  reusing parent's data entry: fullpotential Rulesets
  reusing parent's data entry: fullpotential FormatsData
  reusing parent's data entry: fullpotential Items
  reusing parent's data entry: fullpotential Learnsets
  reusing parent's data entry: fullpotential Moves
  reusing parent's data entry: fullpotential Natures
  reusing parent's data entry: fullpotential Pokedex
  reusing parent's data entry: fullpotential Conditions
  reusing parent's data entry: fullpotential TypeChart
  reusing parent's data entry: fullpotential PokemonGoData
loading data for gen1
loading data for gen2
loading data for gen3
loading data for gen4
loading data for gen5
loading data for gen6
loading data for gen7
loading data for gen8
  reusing parent's data entry: gen8 Natures
  reusing parent's data entry: gen8 Conditions
  reusing parent's data entry: gen8 PokemonGoData
  reusing parent's data entry: gen7 Learnsets
  reusing parent's data entry: gen7 Natures
  reusing parent's data entry: gen7 Conditions
  reusing parent's data entry: gen7 TypeChart
  reusing parent's data entry: gen7 PokemonGoData
  reusing parent's data entry: gen6 Rulesets
  reusing parent's data entry: gen6 Natures
  reusing parent's data entry: gen6 PokemonGoData
  reusing parent's data entry: gen5 Learnsets
  reusing parent's data entry: gen5 Natures
  reusing parent's data entry: gen5 PokemonGoData
  reusing parent's data entry: gen4 Learnsets
  reusing parent's data entry: gen4 Natures
  reusing parent's data entry: gen4 TypeChart
  reusing parent's data entry: gen4 PokemonGoData
  reusing parent's data entry: gen2 Abilities
  reusing parent's data entry: gen2 Natures
loading data for gen1jpn
  reusing parent's data entry: gen1jpn Abilities
  reusing parent's data entry: gen1jpn FormatsData
  reusing parent's data entry: gen1jpn Items
  reusing parent's data entry: gen1jpn Learnsets
  reusing parent's data entry: gen1jpn Natures
  reusing parent's data entry: gen1jpn Pokedex

TypeError: Cannot assign to read only property 'missingno' of object '#<Object>'
    at ModdedDex.modData (redacted/dist/sim/dex.js:169:36)
    at ModdedDex.init (redacted/dist/data/mods/gen1/scripts.js:56:25)
    at ModdedDex.loadData (redacted/dist/sim/dex.js:537:20)
    at ModdedDex.includeData (redacted/dist/sim/dex.js:455:10)
    at ModdedDex.includeModData (redacted/dist/sim/dex.js:450:18)
```

When I add back the `!== Pokedex` check, it works again.

Maybe the bug originates elsewhere, though.

Reply:
To clarify, I tested that on master + `remove Pokedex check`.

Reply:
Weirdly, if I then replace line 433's `return dataObject[dataType];` with `return Object.assign({}, dataObject[dataType]);` it then works again...

Reply:
The bug is with `Scripts.init`. This can be set by inheritance, although that makes no sense...

Reply:
My understanding is that the inherited `Scripts` is only read by line 547, at least during `loadData`, since the `const Scripts` variable on 490 reads a file from `basePath`, and then lines 518 and 552 use that local variable.

Inheriting `Script` in the first place is strange, I agree.

Reply:
aaaand now I get it. Yeah, that's the problem.

Reply:
Indeed, `Scripts` gets `require`d twice, so it's actually the same object as `dataCache.Scripts` all along, and will pick up its copy of the `init` function by inheritance. (Which is why my shallow clone hack worked.)

Reply:
Looks like it got "lost" as part of 2604780.